server/service/src/requisition/request_requisition/insert_program.rs

```diff
@@ -0,0 +1,317 @@
+use crate::{
+    activity_log::activity_log_entry,
+    number::next_number,
+    requisition::{common::check_requisition_exists, query::get_requisition},
+    service_provider::ServiceContext,
+    validate::{check_other_party, CheckOtherPartyType, OtherPartyErrors},
+};
+use chrono::{NaiveDate, Utc};
+use repository::{
+    requisition_row::{RequisitionRow, RequisitionRowStatus, RequisitionRowType},
+    ActivityLogType, NumberRowType, ProgramRequisitionOrderTypeRowRepository,
+    ProgramRequisitionSettingsRowRepository, ProgramRowRepository, RepositoryError, Requisition,
+    RequisitionRowRepository, StorageConnection,
+};
+
+use super::InsertRequestRequisitionError;
+
+#[derive(Debug, PartialEq, Clone, Default)]
+pub struct InsertProgramRequestRequisition {
+    pub id: String,
+    pub other_party_id: String,
+    pub colour: Option<String>,
+    pub their_reference: Option<String>,
+    pub comment: Option<String>,
+    pub expected_delivery_date: Option<NaiveDate>,
+    pub program_order_type_id: String,
+    pub period_id: String,
+}
+
+type OutError = InsertRequestRequisitionError;
+
+pub fn insert_program_request_requisition(
+    ctx: &ServiceContext,
+    input: InsertProgramRequestRequisition,
+) -> Result<Requisition, OutError> {
+    let requisition = ctx
+        .connection
+        .transaction_sync(|connection| {
+            validate(connection, &ctx.store_id, &input)?;
+            let new_requisition = generate(connection, &ctx.store_id, &ctx.user_id, input)?;
+            RequisitionRowRepository::new(&connection).upsert_one(&new_requisition)?;
+
+            activity_log_entry(
+                &ctx,
+                ActivityLogType::RequisitionCreated,
+                Some(new_requisition.id.to_owned()),
+                None,
+            )?;
+
+            get_requisition(ctx, None, &new_requisition.id)
+                .map_err(|error| OutError::DatabaseError(error))?
+                .ok_or(OutError::NewlyCreatedRequisitionDoesNotExist)
+        })
+        .map_err(|error| error.to_inner_error())?;
+
+    Ok(requisition)
+}
+
+fn validate(
+    connection: &StorageConnection,
+    store_id: &str,
+    input: &InsertProgramRequestRequisition,
+) -> Result<(), OutError> {
+    if let Some(_) = check_requisition_exists(connection, &input.id)? {
+        return Err(OutError::RequisitionAlreadyExists);
+    }
+
+    let other_party = check_other_party(
```

Comment:
Sorry, I should have explained a little bit further, this should technically be verified with:

https://github.com/openmsupply/open-msupply/blob/692b37d560cb2f495f3b2d38c86e2310f11b03f2/server/service/src/requisition/program_settings/mod.rs#L31

against: https://github.com/openmsupply/open-msupply/blob/692b37d560cb2f495f3b2d38c86e2310f11b03f2/server/service/src/requisition/program_settings/mod.rs#L23

Reply:
Also order_type and program_id would be checked against that struct, i see you have a todo for that, thanks !